core/shared/src/main/scala/zio/stm/ZSTM.scala

```diff
@@ -161,6 +175,18 @@ final class ZSTM[-R, +E, +A] private[stm] (
   def commitEither: ZIO[R, E, A] =
     either.commit.absolve
 
+  /**
+   * Repeats this `STM` effect until its result satisfies the specified predicate.
+   */
+  final def doUntil(f: A => Boolean): ZSTM[R, E, A] =
+    flatMap(a => if (!f(a)) doUntil(f) else ZSTM.succeedNow(a))
```

Comment:
Could switch the order here if you wanted to avoid the `(!(f(a))`. If the condition is true return immediately, otherwise continue.